src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala

```diff
@@ -229,21 +247,27 @@ class LeonardoService(protected val dataprocConfig: DataprocConfig,
     }
   }
 
-  private[service] def generateServiceAccountKey(googleProject: GoogleProject, serviceAccountEmail: WorkbenchEmail): Future[Option[ServiceAccountKey]] = {
-    // Only generate a key if NOT creating the cluster as the pet service account.
-    // If the pet service account is used to create a cluster, its credentials are on the metadata
-    // server and we don't need to propagate a key file.
-    if (!dataprocConfig.createClusterAsPetServiceAccount) {
-      googleIamDAO.createServiceAccountKey(dataprocConfig.leoGoogleProject, serviceAccountEmail).map(Option(_))
-    } else Future.successful(None)
+  private[service] def generateServiceAccountKey(googleProject: GoogleProject, serviceAccountOpt: Option[WorkbenchEmail]): Future[Option[ServiceAccountKey]] = {
+    serviceAccountOpt.traverse { serviceAccountEmail =>
+      googleIamDAO.createServiceAccountKey(dataprocConfig.leoGoogleProject, serviceAccountEmail)
+    }
+  }
+
+  private[service] def removeServiceAccountKey(googleProject: GoogleProject, clusterName: ClusterName, serviceAccountOpt: Option[WorkbenchEmail]): Future[Unit] = {
+    // Delete the service account key in Google, if present
+    val tea = for {
+      key <- OptionT(dbRef.inTransaction { _.clusterQuery.getServiceAccountKeyId(googleProject, clusterName) })
+      serviceAccountEmail <- OptionT.fromOption[Future](serviceAccountOpt)
+      _ <- OptionT.liftF(googleIamDAO.removeServiceAccountKey(dataprocConfig.leoGoogleProject, serviceAccountEmail, key))
```

Comment:
this might return a Googly exception and nix the " `internalDeleteCluster` must march on regardless of failures" condition

Reply:
though on further thought i'm conflicted about this. we don't monitor the result of the `internalDeleteCluster` future in some cases... do we need delete-retries?

Reply:
(possibly a problem for another PR)

Reply:
That's a good catch. I'm not sure what the right thing to do is. I see the value in having delete march on regardless of intermediate failures. But I also don't like swallowing errors if there is a problem.

Note the call to `googleIamDAO.removeServiceAccountKey` itself retries on errors (and handles 404s): https://github.com/broadinstitute/workbench-libs/blob/develop/google/src/main/scala/org/broadinstitute/dsde/workbench/google/HttpGoogleIamDAO.scala#L168

Is that a reasonable compromise? Or should I jam a recover on this?

Reply:
I'd say we jam a recover on it. We're going to swallow the error anyway; might as well not bail on cleaning up everything else too.

Reply:
I did this in the `internalDeleteCluster` method